README.md

```diff
@@ -261,6 +261,52 @@ are not there.
 
 This makes laziness work right.
 
+### handling end, error, and abort.
+
+in pull streams, any part of the stream (source, sink, or through)
+may terminate the stream. (this is the case with node streams too,
+but it's not handled well).
+
+#### source: end, error
+
+A source may end (`cb(true)` after read) or error (`cb(error)` after read)
+After ending, the source *must* never `cb(null, data)`
+
+#### sink: abort
+
+Sinks do not normally end the stream, but if they decide they do
+not need any more data they may "abort" the source by calling `read(true, cb)`.
+A abort (`read(true, cb)`) may be called before a preceding read call
```

Comment:
yes. you can abort at any time. that is necessary, because if you wait for the preceding read to callback it might never callback and you are hanging open. (this can happen on a real time stream that has stalled)

in the code you have, it should be

``` js
function abortImmediately(read) {
  read(null, cb1) //cb1 should be called before cb2. cb1 may be cb1(err||true)
  read(true, cb2)
}
```